azure/services/bastionhosts/client.go

```diff
@@ -18,139 +18,103 @@ package bastionhosts
-	return &azureClient{c}
-}
-
-// newBastionHostsClient creates a new bastion host client from subscription ID.
-func newBastionHostsClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) network.BastionHostsClient {
-	bastionClient := network.NewBastionHostsClientWithBaseURI(baseURI, subscriptionID)
-	azure.SetAutoRestClientDefaults(&bastionClient.Client, authorizer)
-	return bastionClient
+// newClient creates a new bastion hosts client from an authorizer.
+func newClient(auth azure.Authorizer) (*azureClient, error) {
+	opts, err := azure.ARMClientOptions(auth.CloudEnvironment())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create bastionhosts client options")
+	}
+	factory, err := armnetwork.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create armnetwork client factory")
+	}
+	return &azureClient{factory.NewBastionHostsClient()}, nil
 }
 
 // Get gets the specified bastion host.
 func (ac *azureClient) Get(ctx context.Context, spec azure.ResourceSpecGetter) (result interface{}, err error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "bastionhosts.azureClient.Get")
 	defer done()
 
-	return ac.bastionhosts.Get(ctx, spec.ResourceGroupName(), spec.ResourceName())
+	resp, err := ac.bastionhosts.Get(ctx, spec.ResourceGroupName(), spec.ResourceName(), nil)
+	if err != nil {
+		return nil, err
+	}
+	return resp.BastionHost, nil
 }
 
 // CreateOrUpdateAsync creates or updates a bastion host asynchronously.
-// It sends a PUT request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// It sends a PUT request to Azure and if accepted without error, the func will return a Poller which can be used to track the ongoing
 // progress of the operation.
-func (ac *azureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{}) (result interface{}, future azureautorest.FutureAPI, err error) {
+func (ac *azureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, resumeToken string, parameters interface{}) (result interface{}, poller *runtime.Poller[armnetwork.BastionHostsClientCreateOrUpdateResponse], err error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "bastionhosts.azureClient.CreateOrUpdateAsync")
 	defer done()
 
-	host, ok := parameters.(network.BastionHost)
+	host, ok := parameters.(armnetwork.BastionHost)
 	if !ok {
-		return nil, nil, errors.Errorf("%T is not a network.BastionHost", parameters)
+		return nil, nil, errors.Errorf("%T is not an armnetwork.BastionHost", parameters)
 	}
 
-	createFuture, err := ac.bastionhosts.CreateOrUpdate(ctx, spec.ResourceGroupName(), spec.ResourceName(), host)
+	opts := &armnetwork.BastionHostsClientBeginCreateOrUpdateOptions{ResumeToken: resumeToken}
+	poller, err = ac.bastionhosts.BeginCreateOrUpdate(ctx, spec.ResourceGroupName(), spec.ResourceName(), host, opts)
```

Comment:
I noticed for `natgateways` you added a log statement here (and in `DeleteAsync`). Is that something we want to carry through to all the services? https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/094f41c15f41c871a5d2beabdc229797c883f97b/azure/services/natgateways/client.go#L73-L75

Reply:
I left it out of the other similar PRs because I felt like they were debugging lines I didn't really mean to leave in.